### ISSUE TITLE
HLSL: vertex attribute mapping

### ISSUE THREAD

**Issue by zeux** [KhronosGroup/glslang#643]
We're trying to transition from a Vulkan shader pipeline that involves converting HLSL to Vulkan-compatible GLSL using a separate toolchain (https://github.com/zeux/hlslparser) to glslang. The biggest problem that remains is vertex attribute mapping.

Our vertex shaders generally are of the form

```
struct Appdata
{
    float4 Position	: POSITION;
    float2 Uv	        : TEXCOORD0;
    float3 Normal       : NORMAL;
};

VertexOutput main(Appdata IN) { ... }
```

With attribute locations being defined implicitly by semantics, not by field names (field names are irrelevant under DX9/DX11 binding model). We currently have a fixed mapping (e.g. POSITION -> 0, TEXCOORD0 -> 4) that we use to generate GLSL code in the form of

```
layout(location = 4) in vec2 Uv;
```

I'm trying to figure out how to get anywhere close with glslang building SPIRV from HLSL source directly and failing so far. Things I have discovered:

- Semantic names are fed into `HlslParseContext::handleSemantic`, that converts some built-in semantics into `TQualifier` fields, and discards unknown semantics.
- `TQualifier` has a `layoutLocation` field that *could* be assigned by `HlslParseContext::handleSemantic`, but is later overridden by `HlslParseContext::assignLocations` anyway

I think I understand the data flow but I'm not sure how to proceed. I see several options:

- Fix `HlslParseContext::assignLocations` to preserve existing locations, add custom HLSL syntax to specify attribute locations
- Fix `HlslParseContext::assignLocations` to preserve existing locations, add an option to specify a callback that goes all the way into `HlslParseContext::handleSemantic` and can resolve arbitrary semantic names into locations
- Add an option to specify a callback for `HlslParseContext::assignLocations` to convert qualifier semantics into locations, add TQualifier::semanticName and preserve that in `HlslParseContext::handleSemantic` for unknown semantics
- Anything else?

In general I am curious as to how anybody who works with HLSL & Vulkan handles this right now, since I only see the option of binding by field name - is that what everybody is using currently?

**Comment by sergeyreznik** [comment]
Faced same issue today.

**Comment by sergeyreznik** [comment]
I ended up writing own traverser:
[https://gist.github.com/sergeyreznik/a44eac7316ec085e12e9e4a962ae2147](url)

**Comment by zeux** [comment]
This could be part of a solution I guess, but one other issue that's important - at least for our shaders - is that the *name* of the attribute is generally insufficient, it's the semantics that matters - and the semantics isn't preserved at all from what I can see. For example, we have shaders like this:

```
struct Appdata
{
    int4 Position	: POSITION;

    uint4 Normal	: NORMAL;

    uint4 Material0	: TEXCOORD0;
    uint4 Material1	: TEXCOORD1;
};
```

Where "Material0" is the shader internal name that can change without affecting the C++ code, and "TEXCOORD0" defines the vertex attribute binding. We can convert TEXCOORD0 to an attribute location - this is what we're doing right now for HLSL->MSL translation using hlslparser: https://github.com/Thekla/hlslparser/blob/4221154909ef837b08c882a714dfb64e379b7695/src/MSLGenerator.cpp#L1704-L1743 - but we can't convert "Material0" to an index in a generic way.

**Comment by TiemoJung** [comment]
Hi,

added an interface to control semantics mapping in pull request #660

**Comment by zeux** [comment]
@johnkslang This should probably be tagged as "missing functionality"

**Comment by zeux** [comment]
This should be fixed by #669 - I'll update this issue once this is confirmed.
